File: sdk/rust/src/account/account_balance.rs
/*
 * ‌
 * Hedera Rust SDK
 * ​
 * Copyright (C) 2022 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

use std::collections::HashMap;

use hedera_proto::services;
use prost::Message;

use crate::protobuf::ToProtobuf;
use crate::{
    AccountId,
    FromProtobuf,
    Hbar,
    Tinybar,
    TokenId,
};

/// Response from [`AccountBalanceQuery`][crate::AccountBalanceQuery].
#[derive(Debug, Clone)]
#[cfg_attr(feature = "ffi", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "ffi", serde(rename_all = "camelCase"))]
pub struct AccountBalance {
    /// The account that is being referenced.
    pub account_id: AccountId,

    /// Current balance of the referenced account.
    pub hbars: Hbar,

    /// Token balances for the referenced account.
    #[deprecated = "use a mirror query"]
    #[allow(deprecated)]
    pub tokens: HashMap<TokenId, u64>,

    /// Token decimals for the referenced account.
    #[deprecated = "use a mirror query"]
    #[allow(deprecated)]
    pub token_decimals: HashMap<TokenId, u32>,
}

impl AccountBalance {
    /// Create a new `AccountBalance` from protobuf-encoded `bytes`.
    ///
    /// # Errors
    /// - [`Error::FromProtobuf`](crate::Error::FromProtobuf) if decoding the bytes fails to produce a valid protobuf.
    /// - [`Error::FromProtobuf`](crate::Error::FromProtobuf) if decoding the protobuf fails.
    pub fn from_bytes(bytes: &[u8]) -> crate::Result<Self> {
        FromProtobuf::<services::CryptoGetAccountBalanceResponse>::from_bytes(bytes)
    }

    /// Convert `self` to a protobuf-encoded [`Vec<u8>`].
    #[must_use]
    pub fn to_bytes(&self) -> Vec<u8> {
        #[allow(deprecated)]
        services::CryptoGetAccountBalanceResponse {
            header: None,
            account_id: Some(self.account_id.to_protobuf()),
            balance: self.hbars.to_tinybars() as u64,
            token_balances: Vec::default(),
        }
        .encode_to_vec()
    }
}

impl FromProtobuf<services::CryptoGetAccountBalanceResponse> for AccountBalance {
    #[allow(deprecated)]
    fn from_protobuf(pb: services::CryptoGetAccountBalanceResponse) -> crate::Result<Self>
    where
        Self: Sized,
    {
        let account_id = pb_getf!(pb, account_id)?;
        let account_id = AccountId::from_protobuf(account_id)?;

        let balance = Hbar::from_tinybars(pb.balance as Tinybar);

        let mut tokens = HashMap::with_capacity(pb.token_balances.len());
        let mut token_decimals = HashMap::with_capacity(pb.token_balances.len());

        for token in pb.token_balances {
            let token_id = TokenId::from_protobuf(pb_getf!(token, token_id)?)?;

            tokens.insert(token_id, token.balance);
            token_decimals.insert(token_id, token.decimals);
        }

        Ok(Self { account_id, hbars: balance, tokens, token_decimals })
    }
}

impl FromProtobuf<services::response::Response> for AccountBalance {
    fn from_protobuf(pb: services::response::Response) -> crate::Result<Self> {
        let response = pb_getv!(pb, CryptogetAccountBalance, services::response::Response);

        Self::from_protobuf(response)
    }
}

#[cfg(test)]
mod tests {
    #[cfg(feature = "ffi")]
    mod ffi {
        use std::collections::HashMap;

        use crate::{
            AccountBalance,
            Hbar,
        };

        #[test]
        #[allow(deprecated)]
        fn serialize() {
            expect_test::expect![[r#"
                {
                  "accountId": "0.0.3",
                  "hbars": 200000000,
                  "tokens": {},
                  "tokenDecimals": {}
                }"#]].assert_eq(
                &serde_json::to_string_pretty(&AccountBalance {
                    account_id: 3.into(),
                    hbars: Hbar::new(2),
                    tokens: HashMap::new(),
                    token_decimals: HashMap::new(),
                })
                .unwrap(),
            )
        }
    }
}

File: sdk/swift/Sources/Hedera/Account/AccountBalance.swift
/*
 * ‌
 * Hedera Swift SDK
 * ​
 * Copyright (C) 2022 - 2023 Hedera Hashgraph, LLC
 * ​
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ‍
 */

import Foundation
import HederaProtobufs

private struct TokenBalance {
    fileprivate let id: TokenId
    fileprivate let balance: UInt64
    fileprivate let decimals: UInt32

    internal init(id: TokenId, balance: UInt64, decimals: UInt32) {
        self.id = id
        self.balance = balance
        self.decimals = decimals
    }
}

extension TokenBalance: ProtobufCodable {
    internal typealias Protobuf = Proto_TokenBalance

    internal init(protobuf proto: Protobuf) {
        self.init(id: .fromProtobuf(proto.tokenID), balance: proto.balance, decimals: proto.decimals)
    }

    internal func toProtobuf() -> Protobuf {
        .with { proto in
            proto.tokenID = id.toProtobuf()
            proto.balance = balance
            proto.decimals = decimals
        }
    }
}

/// Response from ``AccountBalanceQuery``.
public struct AccountBalance {
    /// The account that is being referenced.
    public let accountId: AccountId

    /// Current balance of the referenced account.
    public let hbars: Hbar

    private let tokensInner: [TokenBalance]

    // hack to work around deprecated warning
    private var tokenBalancesInner: [TokenId: UInt64] {
        Dictionary(uniqueKeysWithValues: tokensInner.map { ($0.id, $0.balance) })
    }

    /// Token balances for the referenced account.
    ///
    /// This access is *`O(n)`*.
    @available(*, deprecated, message: "use a mirror query")
    public var tokenBalances: [TokenId: UInt64] { tokenBalancesInner }

    // hack to work around deprecated warning
    private var tokenDecimalsInner: [TokenId: UInt32] {
        Dictionary(uniqueKeysWithValues: tokensInner.map { ($0.id, $0.decimals) })
    }

    /// Token decimals for the referenced account.
    ///
    /// This access is *`O(n)`*.
    @available(*, deprecated, message: "use a mirror query")
    public var tokenDecimals: [TokenId: UInt32] { tokenDecimalsInner }

    public static func fromBytes(_ bytes: Data) throws -> Self {
        try Self(protobufBytes: bytes)
    }

    public func toBytes() -> Data {
        self.toProtobufBytes()
    }

    public func toString() -> String {
        String(describing: self)
    }
}

extension AccountBalance: TryProtobufCodable {
    internal typealias Protobuf = Proto_CryptoGetAccountBalanceResponse

    internal init(protobuf proto: Protobuf) throws {
        self.init(
            accountId: try .fromProtobuf(proto.accountID),
            hbars: .fromTinybars(Int64(proto.balance)),
            tokensInner: .fromProtobuf(proto.tokenBalances)
        )
    }

    internal func toProtobuf() -> Protobuf {
        .with { proto in
            proto.accountID = accountId.toProtobuf()
            proto.balance = UInt64(hbars.toTinybars())
            proto.tokenBalances = tokensInner.toProtobuf()
        }
    }
}
